### PR TITLE
Fix orion data models

### LINKED ISSUE #44: Data model stored in Context Broker is wrong
Once you update to Orion >= 0.28 data model needs to be fixed: Here is the data model in normalized mode:

When creating data, if you issue a POST to /v2/entities with the JSON described below it will create the entities correctly. 

``` js
{
    "address": {
        "type": "PostalAddress",
        "value": {
            "addressLocality": "Alegr\u00eda-Dulantzi",
            "addressRegion": "Araba",
            "postalCode": "01240",
            "streetAddress": "Cuesta de las Cabras Aldapa 2"
        }
    },
    "aggregateRating": {
        "value": {
            "ratingValue": 3,
            "reviewCount": 1
        }
    },
    "capacity": {
        "value": 120
    },
    "department": {
        "value": "Franchise3"
    },
    "description": {
        "value": "Restaurante de estilo sidrer\u00eda ubicado en Alegria-Dulantzi. Adem\u00e1s de su men\u00fa del d\u00eda y carta, tambi\u00e9n ofrece men\u00fa de sidrer\u00eda. El men\u00fa del d\u00eda cuesta 9 euros. Los fines de semana la especialidad de la casa son las alubias con sacramentos. En lo que a bebidas se refiere, hay una amplia selecci\u00f3n adem\u00e1s de la sidra. Cabe destacar que se puede hacer txotx. La capacidad del establecimiento es de 50 personas pero la sidrer\u00eda no dispone de aparcamiento.%5cn%5cnHORARIO: %5cn%5cnLunes a domingo: 9:00-17:00 y 19:00-23:00."
    },
    "id": "Elizalde",
    "occupancyLevels": {
        "metadata": {
             "timestamp": {
                "value": 1454350000000.0
            }
        },
        "value": 0
    },
   "location": {
        "type": "geo:point",
        "value": "42.8404625, -2.5123277"
    },
    "priceRange": {
        "value": 0
    },
    "telephone": {
        "value": "945 400 868"
    },
    "type": "Restaurant"
}

```

### PR DESCRIPTION
**Pull request by albertinisg**
Fixes #44 
This PR updates the following:
- Restaurant, review and reservation feeders
- Restaurant, review and reservation input models in the server
- Use 'keyValues' option by default retrieving information from Orion
- Update datetime format to ISO 8601
- Strip/convert forbidden characters from entities id's
- Restaurant, review and reservation models from Orion adapted to Schema.org to not break the client

### CODE REVIEW

File: server/auth/authrequest.js [hunk 1]
```diff
@@ -54,9 +54,12 @@ function performRequest(endpoint, method, data, fiwareHeaders) {
       resolveWithFullResponse: true,
       json: true // Automatically stringifies the body to JSON
     };
+    if (typeof payload !== 'undefined') {
+      endpoint += '?' + querystring.stringify(payload);
+      options.uri = host + ':' + port + endpoint;
```
Comment (jmcanterafonseca) [inline server/auth/authrequest.js:59]: why are these changes needed?

File: server/feeders/reservationsgenerator.js [hunk 2]
```diff
@@ -56,29 +55,38 @@ var feedOrionReservations = function() {
...
+      'reservationStatus': {
+        'value': utils.randomElement(reservations)
+      },
+      'underName': {
+        'type': 'Person',
+        'value': 'user' + utils.randomIntInc(1, 10)
+      },
+      'reservationFor': {
+        'type': 'FoodEstablishment',
+        'value': utils.fixedEncodeURIComponent(restaurantsData[pos].name)
```
Comment (jmcanterafonseca) [inline server/feeders/reservationsgenerator.js:75]: shouldn't this be the restaurant id?

I think we are encoding because Orion does not accept certain chars as value, right?

File: server/feeders/reservationsgenerator.js [hunk 3]
```diff
@@ -56,29 +55,38 @@ var feedOrionReservations = function() {
   };
 
   Object.keys(restaurantsData).forEach(function(element, pos) {
```
Comment (jmcanterafonseca) [inline server/feeders/reservationsgenerator.js:57]: I think it would be better to name the second argument as 'index' instead of 'pos'

File: server/feeders/restaurantfeeder.js [hunk 4]
```diff
@@ -38,12 +38,12 @@ var fiwareHeaders = {
...
-    address += restaurant.address.addressLocality + ' ';
+  var address = restaurant.address.value.streetAddress + ' ';
+  if (restaurant.address.value.addressLocality) {
+    address += restaurant.address.value.addressLocality + ' ';
   }
-  if (restaurant.address.addressRegion) {
-    address += restaurant.address.addressRegion;
+  if (restaurant.address.value.addressRegion) {
+    address += restaurant.address.value.addressRegion;
   }
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:47]: why are we serializing the address to a string?

File: server/feeders/restaurantfeeder.js [hunk 5]
```diff
@@ -74,8 +74,8 @@ var feedOrionRestaurants = function() {
     .catch(function(err) {
       if (err.statusCode == '404') {
         var address = getAddress(attributes);
-        if (typeof attributes.department !== 'undefined') {
-          fwHeaders['fiware-servicepath'] = '/' + attributes.department;
+        if (typeof attributes.department.value !== 'undefined') {
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:77]: if (attributes.department.value) would suffice as I guess department is a string

File: server/feeders/restaurantfeeder.js [hunk 6]
```diff
@@ -74,8 +74,8 @@ var feedOrionRestaurants = function() {
...
+          fwHeaders['fiware-servicepath'] = '/' + attributes.department.value;
         }
         setTimeout(function() {
          geocoder.geocode({address: address, country: 'Spain'})
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:81]: why it is needed a setTimeout here for doing this task? strange, strange

and it should be process.nextTick() why I have a reasonable doubt about its necessity

File: server/feeders/restaurantfeeder.js [hunk 7]
```diff
@@ -132,28 +132,39 @@ var feedOrionRestaurants = function() {
...
-      'department': utils.randomElement(organization),
       'capacity': {
         'type': 'PropertyValue',
-        'name': 'capacity',
         'value': utils.randomElement(capacity)
       },
-      'aggregateRating': {},
+      'aggregateRating': {
+        'type': 'AggregateRating',
+        'value': {}
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:152]: could it be just null? or even not specifying it?

File: server/feeders/restaurantfeeder.js [hunk 8]
```diff
@@ -132,28 +132,39 @@ var feedOrionRestaurants = function() {
...
-        'timestamp': new Date().getTime(),
-        'name': 'occupancyLevels',
         'value': 0
       }
     };
 
-    attr.aggregateRating.ratingValue = utils.randomIntInc(1, 5);
-    attr.aggregateRating.reviewCount = utils.randomIntInc(1,
+    attr.aggregateRating.value.ratingValue = utils.randomIntInc(1, 5);
+    attr.aggregateRating.value.reviewCount = utils.randomIntInc(1,
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:167]: oh you need it here to add this values, ok

File: server/feeders/reviewsgenerator.js [hunk 9]
```diff
@@ -107,16 +109,24 @@ var triggerRestaurantsRatings = function() {
   var servicePath;
 
   var q = async.queue(function(task, callback) {
-
+    var restaurantId = utils.asciiEncode(
+      utils.stripForbiddenChars(
+        unescape(task.restaurantName)));
     setTimeout(function() {
```
Comment (jmcanterafonseca) [inline server/feeders/reviewsgenerator.js:115]: strange strange strange

File: server/feeders/reviewsgenerator.js [hunk 10]
```diff
@@ -107,16 +109,24 @@ var triggerRestaurantsRatings = function() {
   var servicePath;
 
   var q = async.queue(function(task, callback) {
-
+    var restaurantId = utils.asciiEncode(
+      utils.stripForbiddenChars(
+        unescape(task.restaurantName)));
```
Comment (jmcanterafonseca) [inline server/feeders/reviewsgenerator.js:114]: really need to unscape?

File: server/feeders/reviewsgenerator.js [hunk 11]
```diff
@@ -107,16 +109,24 @@ var triggerRestaurantsRatings = function() {
...
+        fiwareHeaders)
       .then(function(data) {
         var fwHeaders = JSON.parse(JSON.stringify(fiwareHeaders));
         if (typeof data.body.department !== 'undefined') {
           fwHeaders['fiware-servicepath'] = '/' + data.body.department;
         }
-        utils.sendRequest('PATCH', task.aggregateRatings,
-          task.restaurantName, fwHeaders)
+        utils.sendRequest(
+          'PATCH',
```
Comment (jmcanterafonseca) [inline server/feeders/reviewsgenerator.js:126]: why are you sending here a patch request?

File: server/utils.js [hunk 12]
```diff
@@ -752,14 +796,49 @@ function getTimeBetweenDates(from, to) {
...
+      },
       'type': 'PropertyValue',
-      'timestamp': date,
-      'name': 'occupancyLevels',
       'value': occupancyLevel
     }
   };
 }
 
+function asciiEncode(string) {
```
Comment (jmcanterafonseca) [inline server/utils.js:811]: is this really necessary?

encodeURIComponent can do the work and less code to maintain

a big -1 to include this kind of stuff

File: server/auth/authrequest.js (hunk 1 above)
Comment (albertinisg) [inline server/auth/authrequest.js:59]: I've added this to allow passing the `keyValues` option when doing POST or PATCH. Before the model change and Orion update, the client was using this format to PATCH the elements so, we need to use `keyValues` to not break the compatibility

**Comment by jmcanterafonseca** [comment]
in general I would avoid such big pull requests in the future. It is a bad practice both for who develop and for who has to review the code. 

Focus your PR in one aspect and made progressive refinement until you reach to where you want to go. 

I was expecting a simple pull request that adapts the bootstrapping generation process but not the whole app

File: server/feeders/restaurantfeeder.js [hunk 13]
```diff
@@ -177,16 +189,19 @@ var feedOrionRestaurants = function() {
...
+            element = utils.fixedEncodeURIComponent(
+              utils.replaceOnceUsingDictionary(
               dictionary, element,
               function(key, dictionary) {
                 return dictionary[key];
-              });
+              }));
             if (element == 'priceRange') {
-              attr[utils.fixedEncodeURIComponent(element)] =
+              attr[element] = {};
```
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:199]: nit:

``` js
attr[element] = {
   value: parseFloat(val)
}
```

shorter

File: server/feeders/reservationsgenerator.js (hunk 2 above)
Comment (albertinisg) [inline server/feeders/reservationsgenerator.js:75]: Yes, it does not accept certain characters. Because of that, for example, white spaces and some other characters are replaced by underscore, and with this conversion we cannot generate the original name again. Using the name allow us to use `name` or `id` just applying the conversion as needed.

File: server/feeders/reviewsgenerator.js [hunk 14]
```diff
@@ -137,7 +146,7 @@ var triggerRestaurantsRatings = function() {
   .then(function(data) {
     reviews = data.body;
     Object.keys(restaurantsData).forEach(function(element, pos) {
-      var restaurantName = restaurantsData[pos].id;
```
Comment (jmcanterafonseca) [inline server/feeders/reviewsgenerator.js:140]: nit: use index

File: server/feeders/restaurantfeeder.js (hunk 4 above)
Comment (albertinisg) [inline server/feeders/restaurantfeeder.js:47]: We adapt it to the format needed for querying the `geocoder` library, to retrieve the geo-location of the restaurant

File: server/feeders/reservationsgenerator.js (hunk 3 above)
Comment (albertinisg) [inline server/feeders/reservationsgenerator.js:57]: I'll take note :)

File: server/routes/orion.js [hunk 15]
```diff
@@ -163,10 +164,9 @@ exports.readRestaurantWithDate = function(req, res) {
       actualOccupancyLevels = utils.getOccupancyLevels(
         restaurantReservations,
         req.params.id);
-      console.log(actualOccupancyLevels);
       occupancyLevelsObject = utils.updateOccupancyLevels(
         actualOccupancyLevels, req.params.date);
-      utils.getListByType('Restaurant', req.params.id, req.headers)
+      utils.getListByType('Restaurant', restaurantId, req.headers)
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:169]: I think it could be too late but indenting every nested promise it is unnatural, because promises were invented just for avoiding nesting of indenting and all that kind of things that make more difficult to follow the code

``` js
a.then(function(x) {
  return b();
}).then(function(y) {
   doWhatever();
})
```

File: server/feeders/restaurantfeeder.js (hunk 5 above)
Comment (albertinisg) [inline server/feeders/restaurantfeeder.js:77]: Ok, so I guess this should be changed in some other places

File: server/routes/orion.js [hunk 16]
```diff
@@ -274,7 +281,7 @@ exports.createReview = function(req, res) {
     if (typeof fwHeaders['fiware-servicepath'] !== 'undefined') {
       delete fwHeaders['fiware-servicepath'];
     }
-    utils.getListByType('Restaurant', restaurantName, fwHeaders)
+    utils.getListByType('Restaurant', restaurantId, fwHeaders)
     .then(function(data) {
       auth.getUserDataPromise(req)
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:286]: I think a return is needed there

File: server/routes/orion.js [hunk 17]
```diff
@@ -376,8 +387,8 @@ exports.updateReview = function(req, res) {
               fixedReviewObject.reviewBody = utils.fixedEncodeURIComponent(
                 fixedReviewObject.reviewBody);
             }
-            utils.sendRequest('PATCH', fixedReviewObject, req.params.id,
-                req.headers)
+            utils.sendRequest('PATCH', fixedReviewObject, reviewId,
+                req.headers, {'options': 'keyValues'})
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:391]: too much nesting here, consider separate into functions and to create a more readable structure taking advantage of promises

File: server/routes/orion.js [hunk 18]
```diff
@@ -458,7 +472,7 @@ exports.deleteReview = function(req, res) {
                     restaurantReviews);
                   req.headers['fiware-servicepath'] = '/' + servicePath;
                   utils.sendRequest('PATCH', aggregateRatings,
-                      restaurantName, req.headers)
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:461]: too much nesting, refactor

File: server/routes/orion.js [hunk 19]
```diff
@@ -564,15 +578,16 @@ exports.createReservation = function(req, res) {
...
-        req.body.reservationFor.name,
+        restaurantId,
         req.headers)
     .then(function(data) {
       elementToOrion = req.body;
-      elementToOrion.reservationFor.address = data.body.address;
+      elementToOrion.address = data.body.address;
       capacity = data.body.capacity.value;
       timeframeQuery = utils.getTimeframe(req.body.startTime);
       utils.sendRequest('GET', {
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:593]: refactor and make better use of promises

File: server/routes/orion.js [hunk 20]
```diff
@@ -677,7 +694,11 @@ exports.updateReservation = function(req, res) {
...
         });
       } else {
-        utils.sendRequest('PATCH', req.body, req.params.id, req.headers)
+        utils.sendRequest('PATCH',
+                          req.body,
+                          reservationId,
+                          req.headers,
+                          {'options': 'keyValues'})
         .then(function(data) {
           res.statusCode = data.statusCode;
```
Comment (jmcanterafonseca) [inline server/routes/orion.js:703]: same comment

File: server/feeders/restaurantfeeder.js (hunk 6 above)
Comment (albertinisg) [inline server/feeders/restaurantfeeder.js:81]: I've added this `setTimeout` to wait for the `geocoder` to return the desired object. I did not find any better approach at that time. I'll have a look to the `process.nextTick()`. Thanks!

File: server/feeders/restaurantfeeder.js (hunk 7 above)
Comment (albertinisg) [inline server/feeders/restaurantfeeder.js:152]: If I do not specify it, it returns an `undefined` error when doing `attr.aggregateRating.value.ratingValue`. Maybe null could work. (I'll have a look)

File: server/auth/authrequest.js (hunk 1 above)
Comment (jmcanterafonseca) [inline server/auth/authrequest.js:59]: then name it better to understand it is an option not a payload and in all case it is a queryString not a payload

File: server/utils.js [hunk 21]
```diff
@@ -632,36 +678,51 @@ function getOrgReservations(franchise, listOfRestaurants, listOfReservations) {
...
+      headers
+    );
+  } else {
+    return authRequest(
+      uri,
+      method,
+      body,
+      headers,
+      payload
+    );
```
Comment (jmcanterafonseca) [inline server/utils.js:724]: strange, strange

simply check the payload argument at the called function and you will save precious lines of code

File: server/utils.js [hunk 22]
```diff
@@ -632,36 +678,51 @@ function getOrgReservations(franchise, listOfRestaurants, listOfReservations) {
...
 
-function getListByType(type, element, headers) {
+function getListByType(type, element, headers, raw) {
   var uri = '/v2/entities';
+  var options = {'type': type,'limit': '1000'};
   if (element) {
     uri += '/' + encodeURIComponent(element);
   }
+
+  if (typeof raw === 'undefined') {
```
Comment (jmcanterafonseca) [inline server/utils.js:694]: raw,

why don't you name the param keyValues and pass a true or false ?

it will be clearer

File: server/utils.js [hunk 23]
```diff
@@ -454,49 +426,54 @@ function dataToSchema(listOfElements) {
...
 
 function addGeolocation(schemaObject, geoObject) {
+  // The returned object will be POST/PATCH(ed),
+  // so we need to add the 'value' field
   if (geoObject) {
-    schemaObject.position = {};
-    schemaObject.position.type = 'coords';
-    schemaObject.position.location = 'WGS84';
-    schemaObject.position.value = geoObject.latitude + ', ' +
+    schemaObject.location = {};
```
Comment (jmcanterafonseca) [inline server/utils.js:474]: nit: 

``` js
schemaObject.location = {
  type: 'geo:point',
  value: 'xx, yy' 
}
```

shorter

File: server/utils.js [hunk 24]
```diff
@@ -454,49 +426,54 @@ function dataToSchema(listOfElements) {
   listOfElements = objectToArray(listOfElements);
 
   Object.keys(listOfElements).forEach(function(element, pos) {
-
-    newElement = objectDataToSchema(listOfElements[pos]);
+    if (typeof date !== 'undefined') {
```
Comment (jmcanterafonseca) [inline server/utils.js:429]: same comment do that checking on the called function and not here you will make the code easier to read

File: server/utils.js [hunk 25]
```diff
@@ -408,7 +376,11 @@ function objectDataToSchema(element) {
 
   case 'Restaurant':
 
-    newElement = restaurantToSchema(element);
+    if (typeof date !== 'undefined') {
```
Comment (jmcanterafonseca) [inline server/utils.js:379]: same comment as below

File: server/feeders/restaurantfeeder.js (hunk 6 above)
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:81]: use promises, run the geocoder and then continue,

**Comment by jmcanterafonseca** [comment]
thanks Alberto for the good work

we need several improvements, including refactorings, to this and I put under your consideration to refactor this PR into smaller pieces. 

piece 1: the code that feeds orion with the right data. I think we are close to this and that code could land soon. 
piece 2: the application logic that operates over that data. we are far because some refactoring is needed 

do you agree with this approach?

thanks

**Comment by albertinisg** [comment]
I'm agree. I'll let this PR open until the new modifications are done. I'll let the 'piece 1' for this sprint, and the next one for the next month (as it will need more refactoring).

Thanks!

File: server/feeders/restaurantfeeder.js (hunk 6 above)
Comment (albertinisg) [inline server/feeders/restaurantfeeder.js:81]: I've removed it and now I remember why I've used the `setTimeout` function. When doing several requests against the Google maps API, it allows to do 10 requests per second. Sometimes this limit is reached and, we get then the error `[Error: Status is OVER_QUERY_LIMIT. You have exceeded your rate-limit for this API.]`. The only different thing we can do here to avoid the error is, catch it, and then launch the request again if failed; but this can lead into an infinite loop when the 2500 requests per day are reached, as the error returned is the same. So the only thing now it comes to my mind is keep the `setTimeout`... Do you think there is actually a better approach for it? Suggestions are welcome :)

File: server/feeders/reviewsgenerator.js (hunk 11 above)
Comment (albertinisg) [inline server/feeders/reviewsgenerator.js:126]: As we are generating reviews, we need to update the restaurant Ratings. So every time a review is added, it recalculates the rating and updates the field in the Restaurant entity.

File: server/feeders/reviewsgenerator.js (hunk 10 above)
Comment (albertinisg) [inline server/feeders/reviewsgenerator.js:114]: I've gone crazy with this 'encoding' thing. I've tested everything, and that was the only way I could perform all the requests (for the 688 restaurants). Here are my tries:
- Using the following code:

``` js
var restaurantId = utils.stripForbiddenChars(task.restaurantName);
```

It queries (URI encoded):

```
http://orion:1026/v2/entities/Se%C3%B1orio_de_Astobiza_%2528Okendo_Txakolina%2529
```

Which returns:

``` js
{ error: 'BadRequest', description: 'invalid character in URI' }
```

Here's the first issue: the `ñ` (note that this behaviour is the same for `á, é, í` and so on). As stands in the specification, field sintax restrictions, it says 'Allowed characters are the ones in the plain ASCII set'. Then I made the function asciiEncode and try again:

``` js
var restaurantId = utils.asciiEncode(utils.stripForbiddenChars(task.restaurantName));
```

This queries:

```
http://orion:1026/v2/entities/Se%5Cu00f1orio_de_Astobiza_%2528Okendo_Txakolina%2529
```

And returns the following:

``` js
{ error: 'NotFound',
  description: 'The requested entity has not been found. Check type and id' }
```

This is because, the entity id is loaded in Orion as `Señorio_de_Astobiza__Okendo_Txakolina_`, but the name, the variable `task.restaurantName` is the name attribute of that id stored in Orion. The thing is, to store the original restaurant name before, we had to URI encode it (name), so the name is stored in Orion entity as `Señorio de Astobiza %28Okendo Txakolina%29`, otherwise it returns the following error:

``` js
{ error: 'BadRequest',
        description: 'Invalid characters in attribute value' } } }
```

Due to that, the only thing we could do to accomplish this is:
- Get the name from the Restaurant entity (URI encoded)
- Unescape it (or `decodeURIComponent` should work also but I did not test it)
- Strip the forbidden characters
- Ascii encode the rest of them

With that procedure we can conclude in:

```
var restaurantId = utils.asciiEncode(utils.stripForbiddenChars(unescape(task.restaurantName)));
```

Which does the following request:

```
http://orion:1026/v2/entities/Se%5Cu00f1orio_de_Astobiza__Okendo_Txakolina_
```

And finally returns the entity we want to.

I do know this is a mess and I'm the first one I dislike this procedure, but it was the only way I found to load all the restaurants.

File: server/feeders/restaurantfeeder.js (hunk 6 above)
Comment (jmcanterafonseca) [inline server/feeders/restaurantfeeder.js:81]: that's certainly an issue. I would prefer that we migrate to an open geo coder, and remove Google dependencies. Please could you open an issue in order to fix that?

File: server/feeders/reviewsgenerator.js (hunk 10 above)
Comment (jmcanterafonseca) [inline server/feeders/reviewsgenerator.js:114]: why don't we use an id which it is not the name of the restaurant? you would avoid many troubles. The code is starting to get complex and you should take into account that this is a tutorial ...

**Comment by albertinisg** [comment]
Closing this as it's not needed anymore (all changes regarding this PR are already implemented)
